src/profile-logic/comparison.js

```diff
@@ -0,0 +1,60 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+// @flow
+
+import type { IndexIntoCategoryList, CategoryList } from '../types/profile';
+
+type TranslationMapForCategories = Map<
+  IndexIntoCategoryList,
+  IndexIntoCategoryList
+>;
+export function mergeCategories(
+  categories1: CategoryList,
+  categories2: CategoryList
+): {
+  categories: CategoryList,
+  translationMaps: TranslationMapForCategories[],
+} {
+  const newCategories = [];
+  const translationMaps = [];
+
+  [categories1, categories2].forEach(categories => {
+    const translationMap = new Map();
+    translationMaps.push(translationMap);
+    const insertedCategories: Map<string, IndexIntoCategoryList> = new Map();
+
+    categories.forEach((category, i) => {
+      const { name } = category;
+      const insertedCategoryIndex = insertedCategories.get(name);
+      if (insertedCategoryIndex !== undefined) {
+        translationMap.set(i, insertedCategoryIndex);
+        return;
+      }
+
+      translationMap.set(i, newCategories.length);
+      insertedCategories.set(name, newCategories.length);
+      newCategories.push(category);
+    });
+  });
+
+  return { categories: newCategories, translationMaps };
+}
+
+export function adjustCategories(
+  categories: $ReadOnlyArray<IndexIntoCategoryList | null>,
+  translationMap: TranslationMapForCategories
+): Array<IndexIntoCategoryList | null> {
+  return categories.map(category => {
+    if (category === null) {
+      return null;
+    }
+    const result = translationMap.get(category);
+    if (result === undefined) {
+      throw new Error(
+        `Category with index ${category} wasn't found in the translation map.`
```

Comment:
It's not obvious to me, does this error indicate that:

A. The categories differed between profiles
B. That some internal invariant of how these maps were generated is wrong?

If it's A, then I think we shouldn't throw an error for it, and should gracefully handle it.
If it's B, then perhaps make the error message a bit more precise that this is a program invariant.

Reply:
This is more B. Actually, this was really to make Flow happy :) I'll change the error message as you propose.